### PR TITLE
feat: make elliptic curve arithmetic complete by default and only fallback to incomplete path on request

### PR DESCRIPTION
**Pull request by ivokub**
# Description

This PR makes the elliptic curve arithmetic safe by default, not through an option. In most of the uses cases we can assume adverserial input, so we would like to avoid hitting the incomplete cases which break proof generation. In principle it is a breaking change as we change the defaults, but to the safer less efficient side.

Also fixed a few soundness issues where a malicious prover could prove the statements for invalid inputs.

Also removed incomplete paths in cases where the performance improvements were marginal. This was motivated by the fact that in majority of methods the incomplete set was bigger than initially described (zero scalar or trivial point) and fixing all edge cases would was complicated or very inefficient. We now instead document the set of degenerate cases when we can.

As utility function I added an internal `compilelogger` package which allows to log things only once during circuit compilation time (just to hint the users and not to flood the output). Also update stats package to skip 0-constraint snippets.

See a generated summary below:

### 1. `std/algebra/emulated/sw_emulated`

- Fixed complete-path soundness bugs where hinted scalar-mul outputs could be returned without being fully constrained on edge branches.
  - `scalarMulFakeGLV`: zero-scalar / infinity branches now return canonical infinity instead of unchecked hinted output.
  - `scalarMulGLVAndFakeGLV`: removed the hinted `X`-coordinate bypass and switched edge handling to constrained scalar/point branches with canonical returns.
- Fixed `AddUnified` for the `j=0` branch so the inverse-case selector no longer overrides prior infinity handling when one input is `(0,0)` and the other is a valid `X=0` point.
- Kept incomplete support where the savings are large enough to matter, but documented the real contract.
  - `scalarMulFakeGLV`: incomplete exceptional set documented as
    - `Q = (0,0)`
    - `s = 0`
    - `s ∈ {±1, ±3, ±3⁻¹ mod r}`
  - `scalarMulGLVAndFakeGLV`: incomplete exceptional set documented as
    - `P = (0,0)`
    - `s ∈ {0, ±1}`
    - plus a sparse point-dependent exceptional set from precomputations and the bias step
  - `scalarMulGLV`: exceptional set documented as implementation-dependent
  - `scalarMulJoye`: exceptional set documented as unknown
- Updated wrapper docs so they no longer over-promise simple sufficient conditions.
  - `ScalarMul` now explicitly enumerates endomorphism curves vs non-endomorphism curves.
  - `JointScalarMulBase` and `MultiScalarMul` now describe incomplete mode as inheriting lower-level exceptional sets and additional sparse accumulator/prefix-sum collisions.

### 2. `std/algebra/emulated/sw_bls12381`

- Removed effective public support for `WithIncompleteArithmetic` in the public G2 wrappers.
  - Public `ScalarMul` and `MultiScalarMul` now always use complete arithmetic.
  - If `WithIncompleteArithmetic` is provided, it is ignored and logged at compile time through `compilelogger.LogOnce`.
- Kept private/internal test helpers as-is, but updated their comments:
  - `scalarMulGeneric`: incomplete exceptional set documented as unknown / implementation-dependent
  - `scalarMulGLV`: incomplete exceptional set documented as sparse and point-dependent
- Rationale: the constraint savings were too small to justify a public incomplete path.

### 3. `std/algebra/emulated/sw_bn254` and `std/algebra/emulated/sw_bw6761`

- No public `WithIncompleteArithmetic` scalar-mul API exists there, so no cleanup was required.

### 4. `std/algebra/native/sw_bls12377` G1

- Removed `constScalarMul`.
- Renamed `varScalarMul` to `scalarMulGLV`.
- `ScalarMul` and `ScalarMulBase` now route through `scalarMulGLV`.
- `scalarMulGLV` logs and ignores `WithIncompleteArithmetic`; complete arithmetic is always used there.
- Rationale:
  - the old constant-scalar path had extremely non-uniform cost
  - it was excellent for tiny sparse constants, but much worse for dense large constants
  - removing it simplified the implementation materially
- Kept incomplete support for `MultiScalarMul`, but updated docs to describe its incomplete contract honestly.
- Updated `jointScalarMul` / `jointScalarMulUnsafe` docs to state:
  - complete arithmetic is the default
  - incomplete mode has at least zero-scalar, infinity-input, and `Q = ±R` degeneracies
  - additional sparse implementation-dependent failures may arise from precomputations and accumulator collisions

### 5. `std/algebra/native/sw_grumpkin` G1

- Applied the same simplification as BLS12-377 G1:
  - removed `constScalarMul`
  - renamed `varScalarMul` to `scalarMulGLV`
  - `ScalarMul` / `ScalarMulBase` now route through `scalarMulGLV`
  - `scalarMulGLV` logs and ignores `WithIncompleteArithmetic`
- Kept incomplete support for `MultiScalarMul`, with the same documented caveat that the exceptional set is sparse and implementation-dependent at the API level.
- Updated `jointScalarMul`, `jointScalarMulUnsafe`, and `jointScalarMulGLVUnsafe` docs accordingly.

### 6. `std/algebra/native/twistededwards`

- No `WithIncompleteArithmetic` hook exists in this package.
- No cleanup was needed for this sweep.

### `std/algebra/emulated/sw_emulated`

- `P256 ScalarMul`
  - complete: `328895`
  - incomplete: `263176`
  - reduction: `20.0%`
- `secp256k1 ScalarMul`
  - complete: `426780`
  - incomplete: `202501`
  - reduction: `52.5%`
- `secp256k1 ScalarMulBase`
  - complete: `423729`
  - incomplete: `202000`
  - reduction: `52.3%`
- `secp256k1 JointScalarMulBase`
  - complete: `828881`
  - incomplete: `314875`
  - reduction: `62.0%`
- `secp256k1 MultiScalarMul(2)`
  - complete: `832123`
  - incomplete: `315378`
  - reduction: `62.1%`

Decision:
- keep incomplete support, but weaken/document the contract honestly

### `std/algebra/emulated/sw_bls12381`

- `bls12381 G2 ScalarMul`
  - complete: `717412`
  - incomplete: `709309`
  - reduction: `1.13%`
- `bls12381 G2 MultiScalarMul(2)`
  - complete: `1293825`
  - incomplete: `1276741`
  - reduction: `1.32%`

Decision:
- remove effective public incomplete support

### `std/algebra/native/sw_bls12377`

- `ScalarMul`
  - complete: `4063`
  - incomplete: `3955`
  - reduction: `2.66%`
- `ScalarMulBase`
  - complete: `3248`
  - incomplete: `3185`
  - reduction: `1.94%`
- `MultiScalarMul(2)`
  - complete: `8067`
  - incomplete: `6777`
  - reduction: `15.99%`

Decision:
- remove effective incomplete support for public `ScalarMul` / `ScalarMulBase`
- keep incomplete support for `MultiScalarMul`

### `std/algebra/native/sw_grumpkin`

- `ScalarMul`
  - complete: `5120`
  - incomplete: `5033`
  - reduction: `1.70%`
- `ScalarMulBase`
  - complete: `4302`
  - incomplete: `4260`
  - reduction: `0.98%`
- `MultiScalarMul(2)`
  - complete: `9961`
  - incomplete: `8713`
  - reduction: `12.53%`

Decision:
- remove effective incomplete support for public `ScalarMul` / `ScalarMulBase`
- keep incomplete support for `MultiScalarMul`

### Constant-scalar path measurements that motivated removing `constScalarMul`

Measured on the original pre-change code:

- Small constant `42`
  - `sw_bls12377`: variable `3693`, constant `537` (`85.46%` savings)
  - `sw_grumpkin`: variable `4889`, constant `537` (`89.02%` savings)
- Large constant `modulus - 12345`
  - `sw_bls12377`: variable `3693`, constant `9692` (`162.44%` worse)
  - `sw_grumpkin`: variable `4889`, constant `9247` (`89.14%` worse)

Conclusion:
- the specialized constant path was too bimodal and not worth keeping in the simplified public API
## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

# How has this been tested?

Existing tests.

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [x] `golangci-lint` does not output errors locally
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

---

> [!NOTE]
> **High Risk**
> High risk because it changes elliptic-curve arithmetic semantics and implementations across multiple curve backends (including scalar-mul/add formulas and EVM/KZG verification paths), which can affect proof correctness, constraint counts, and soundness in cryptography-critical code.
> 
> **Overview**
> Makes elliptic-curve operations **complete by default** by introducing `algopts.WithIncompleteArithmetic()` (and deprecating `WithCompleteArithmetic()`), then flipping many scalar-mul/MSM call sites and implementations to treat zero scalars and infinity points as supported inputs.
> 
> Hardens and refactors curve arithmetic in `sw_emulated` and several native curves: fixes `AddUnified` edge cases on j=0 curves, ensures scalar-mul helpers return *constrained canonical outputs* on complete-arithmetic branches (instead of unchecked hint outputs), and adds regression tests for previously exploitable/silent-failure cases.
> 
> Removes/ignores incomplete-arithmetic in some public APIs where gains are marginal (logging once at compile time via new `internal/compilelogger.LogOnce`), updates stats generation to skip missing backend/curve entries, and adjusts EVM/KZG precompile code to rely on complete MSM/scalar-mul behavior (with targeted dummy routing for off-curve branches).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 8ca694a4fc854194917f3734c784f9541b84fe22. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: std/signature/ecdsa/ecdsa.go [hunk 1]
```diff
@@ -63,17 +70,28 @@ func (pk PublicKey[T, S]) prepareVerification(api frontend.API, params sw_emulat
...
+	// Callers use this to either assert (Verify) or mask the result (IsValid).
+	rIsZero := scalarApi.IsZero(&sig.R)
+	sIsZero := scalarApi.IsZero(&sig.S)
 	pkpt := sw_emulated.AffinePoint[T](pk)
+	xIsZero := baseApi.IsZero(&pkpt.X)
+	yIsZero := baseApi.IsZero(&pkpt.Y)
+	pkIsInfinity := api.And(xIsZero, yIsZero)
+	anyInvalid := api.Or(api.Or(rIsZero, sIsZero), pkIsInfinity)
+	inputsValid := api.Sub(1, anyInvalid)
+
```
[the remark this answers is not in the document]
Comment (ivokub) [reply std/signature/ecdsa/ecdsa.go:99]: Fixed

File: std/algebra/emulated/sw_emulated/point.go [hunk 2]
```diff
@@ -234,48 +234,113 @@ func (c *Curve[B, S]) AssertIsOnCurve(p *AffinePoint[B]) {
...
+			X: *xr,
+			Y: *yr,
+		}
+
+		// if p=(0,0) return q
+		result = c.Select(isPInfinity, q, result)
+		// if q=(0,0) return p
+		result = c.Select(isQInfinity, p, result)
+		// if p.y + q.y = 0, return (0, 0)
+		result = c.Select(isYSumZero, infinity, result)
```
[the remark this answers is not in the document]
Comment (ivokub) [reply std/algebra/emulated/sw_emulated/point.go:346]: Correct, but does not apply to supported curves.

File: std/algebra/emulated/sw_emulated/point.go [hunk 3]
```diff
@@ -1229,11 +1356,11 @@ func (c *Curve[B, S]) scalarMulFakeGLV(Q *AffinePoint[B], s *emulated.Element[S]
...
 
-	var selector1 frontend.Variable
+	var isScalarZero frontend.Variable
 	_s := s
-	if cfg.CompleteArithmetic {
-		selector1 = c.scalarApi.IsZero(s)
-		_s = c.scalarApi.Select(selector1, c.scalarApi.One(), s)
+	if !cfg.IncompleteArithmetic {
+		isScalarZero = c.scalarApi.IsZero(s)
+		_s = c.scalarApi.Select(isScalarZero, c.scalarApi.One(), s)
```
[the remark this answers is not in the document]
Comment (ivokub) [reply std/algebra/emulated/sw_emulated/point.go:1363]: Invalid report, we perform the check on lines 1373 and 1381

**Comment by ivokub** [comment]
@yelhousni - should be review now. I think it would be good to merge before doing any other changes to the curves as it is quite a big refactoring over EC gadgets.

**Comment by yelhousni** [comment]
good to merge on my side!
